### ISSUE TITLE
"Unknown License" reported for reusable workflow dependencies (likely API issue)

### ISSUE THREAD

**Issue by marcransome** [actions/dependency-review-action#612]
I'm in the process of migrating to reusable workflows in an effort to reduce duplication across projects, and `dependency-review-action` is reporting "Unknown License" for reusable workflow dependencies as a result, as seen here: https://github.com/fish-shop/install-plugin/pull/31. I suspect this to be an issues with the Dependency Graph API rather than the GitHub Action, and I'm happy to report this elsewhere or in the community forum if needed.

Looking at the Dependency Graph diff for the subset of dependencies being added in the [pull request](https://github.com/fish-shop/install-plugin/pull/31), the `source_repository_url` attributes are `null`:

**Request:**

```shell
gh api repos/fish-shop/install-plugin/dependency-graph/compare/main...aa9db70e981cd281973820c2f48e139a7e1235d0
```

**Response:**

```json
[
  {
    "change_type": "added",
    "manifest": ".github/workflows/dependency-review.yml",
    "ecosystem": "actions",
    "name": "fish-shop/workflows/.github/workflows/dependency-review.yml",
    "version": "1",
    "package_url": "pkg:githubactions/fish-shop/workflows/.github/workflows/dependency-review.yml@1",
    "license": null,
    "source_repository_url": null,
    "scope": "runtime",
    "vulnerabilities": []
  },
  {
    "change_type": "added",
    "manifest": ".github/workflows/release-tags.yml",
    "ecosystem": "actions",
    "name": "fish-shop/workflows/.github/workflows/release-tags.yml",
    "version": "1",
    "package_url": "pkg:githubactions/fish-shop/workflows/.github/workflows/release-tags.yml@1",
    "license": null,
    "source_repository_url": null,
    "scope": "runtime",
    "vulnerabilities": []
  }
```

It looks like the `source_repository_url` value is used by the action as a fallback to determine the license from the Licenses API. The absence of a `source_repository_url` value means there is no string to parse the `owner` and `repo` values from, and no subsequent API request is made to retrieve the license details:

https://github.com/actions/dependency-review-action/blob/fde92acd0840415674c16b39c7d703fc28bc511e/src/licenses.ts#L146-L150

The repository containing the reusable workflows ([fish-shop/workflows](https://github.com/fish-shop/workflows)) does correctly return license details and a `spdx_id` value:

**Request:**

```shell
gh api repos/fish-shop/workflows/license
```

**Response:**

```json
{
  "name": "LICENSE",
  "path": "LICENSE",
  "sha": "bfe650c30cea61def97dc273834f09b95f8b2fc8",
  "size": 1066,
  "url": "https://api.github.com/repos/fish-shop/workflows/contents/LICENSE?ref=main",
  "html_url": "https://github.com/fish-shop/workflows/blob/main/LICENSE",
  "git_url": "https://api.github.com/repos/fish-shop/workflows/git/blobs/bfe650c30cea61def97dc273834f09b95f8b2fc8",
  "download_url": "https://raw.githubusercontent.com/fish-shop/workflows/main/LICENSE",
  "type": "file",
  "content": "TUlUIExpY2Vuc2UKCkNvcHlyaWdodCAoYykgMjAyMyBmaXNoLXNob3AKClBl\ncm1pc3Npb24gaXMgaGVyZWJ5IGdyYW50ZWQsIGZyZWUgb2YgY2hhcmdlLCB0\nbyBhbnkgcGVyc29uIG9idGFpbmluZyBhIGNvcHkKb2YgdGhpcyBzb2Z0d2Fy\nZSBhbmQgYXNzb2NpYXRlZCBkb2N1bWVudGF0aW9uIGZpbGVzICh0aGUgIlNv\nZnR3YXJlIiksIHRvIGRlYWwKaW4gdGhlIFNvZnR3YXJlIHdpdGhvdXQgcmVz\ndHJpY3Rpb24sIGluY2x1ZGluZyB3aXRob3V0IGxpbWl0YXRpb24gdGhlIHJp\nZ2h0cwp0byB1c2UsIGNvcHksIG1vZGlmeSwgbWVyZ2UsIHB1Ymxpc2gsIGRp\nc3RyaWJ1dGUsIHN1YmxpY2Vuc2UsIGFuZC9vciBzZWxsCmNvcGllcyBvZiB0\naGUgU29mdHdhcmUsIGFuZCB0byBwZXJtaXQgcGVyc29ucyB0byB3aG9tIHRo\nZSBTb2Z0d2FyZSBpcwpmdXJuaXNoZWQgdG8gZG8gc28sIHN1YmplY3QgdG8g\ndGhlIGZvbGxvd2luZyBjb25kaXRpb25zOgoKVGhlIGFib3ZlIGNvcHlyaWdo\ndCBub3RpY2UgYW5kIHRoaXMgcGVybWlzc2lvbiBub3RpY2Ugc2hhbGwgYmUg\naW5jbHVkZWQgaW4gYWxsCmNvcGllcyBvciBzdWJzdGFudGlhbCBwb3J0aW9u\ncyBvZiB0aGUgU29mdHdhcmUuCgpUSEUgU09GVFdBUkUgSVMgUFJPVklERUQg\nIkFTIElTIiwgV0lUSE9VVCBXQVJSQU5UWSBPRiBBTlkgS0lORCwgRVhQUkVT\nUyBPUgpJTVBMSUVELCBJTkNMVURJTkcgQlVUIE5PVCBMSU1JVEVEIFRPIFRI\nRSBXQVJSQU5USUVTIE9GIE1FUkNIQU5UQUJJTElUWSwKRklUTkVTUyBGT1Ig\nQSBQQVJUSUNVTEFSIFBVUlBPU0UgQU5EIE5PTklORlJJTkdFTUVOVC4gSU4g\nTk8gRVZFTlQgU0hBTEwgVEhFCkFVVEhPUlMgT1IgQ09QWVJJR0hUIEhPTERF\nUlMgQkUgTElBQkxFIEZPUiBBTlkgQ0xBSU0sIERBTUFHRVMgT1IgT1RIRVIK\nTElBQklMSVRZLCBXSEVUSEVSIElOIEFOIEFDVElPTiBPRiBDT05UUkFDVCwg\nVE9SVCBPUiBPVEhFUldJU0UsIEFSSVNJTkcgRlJPTSwKT1VUIE9GIE9SIElO\nIENPTk5FQ1RJT04gV0lUSCBUSEUgU09GVFdBUkUgT1IgVEhFIFVTRSBPUiBP\nVEhFUiBERUFMSU5HUyBJTiBUSEUKU09GVFdBUkUuCg==\n",
  "encoding": "base64",
  "_links": {
    "self": "https://api.github.com/repos/fish-shop/workflows/contents/LICENSE?ref=main",
    "git": "https://api.github.com/repos/fish-shop/workflows/git/blobs/bfe650c30cea61def97dc273834f09b95f8b2fc8",
    "html": "https://github.com/fish-shop/workflows/blob/main/LICENSE"
  },
  "license": {
    "key": "mit",
    "name": "MIT License",
    "spdx_id": "MIT",
    "url": "https://api.github.com/licenses/mit",
    "node_id": "MDc6TGljZW5zZTEz"
  }
}
```

Any advice would be appreciated. 🥳

**Comment by febuiles** [comment]
@marcransome Thanks for bringing this up! I'm not sure if your use case can be achieved with how Dependency Graph handles GitHub Actions today.

_tl;dr `source_repository_url` can't be set because Dependency Graph is looking for a package named `fish-shop/workflows/.github/workflows/dependency-review.yml`._

I created a sample repo with a very similar change to yours and called `compare` on my PR:

```
$ gh api repos/future-funk/fantastic-invention/dependency-graph/compare/febuiles-patch-1^...febuiles-patch-1
[
  {
    "change_type": "added",
    "manifest": ".github/workflows/dr.yml",
    "ecosystem": "actions",
    "name": "fish-shop/workflows/.github/workflows/dependency-review.yml",
    "version": "1",
    "package_url": "pkg:githubactions/fish-shop/workflows/.github/workflows/dependency-review.yml@1",
    "license": null,
    "source_repository_url": null,
    "scope": "runtime",
    "vulnerabilities": []
  },
  {
    "change_type": "removed",
    "manifest": ".github/workflows/dr.yml",
    "ecosystem": "actions",
    "name": "actions/checkout",
    "version": "4",
    "package_url": "pkg:githubactions/actions/checkout@4",
    "license": null,
    "source_repository_url": "https://github.com/actions/checkout",
    "scope": "runtime",
    "vulnerabilities": []
  },
  {
    "change_type": "removed",
    "manifest": ".github/workflows/dr.yml",
    "ecosystem": "actions",
    "name": "actions/dependency-review-action",
    "version": "3",
    "package_url": "pkg:githubactions/actions/dependency-review-action@3",
    "license": null,
    "source_repository_url": "https://github.com/actions/dependency-review-action",
    "scope": "runtime",
    "vulnerabilities": []
  }
]
```

Note how `name` is valid for the last two, but has something weird for the first one (`fish-shop/workflows/.github/workflows/dependency-review.yml`). When we go from using:

```
- name: Dependency review
        uses: actions/dependency-review-action@v3
```

To:

```
    uses: fish-shop/workflows/.github/workflows/dependency-review.yml@v1
```

Dependency Graph is no longer able to track the Actions inside `dependency-review.yml` because it thinks that everything after `uses` is a package. This is how the Dependency Graph looks for my project after merging the PR:

![Image](https://github.com/actions/dependency-review-action/assets/2161/2b1fa0db-ad76-45b9-aacd-eccca0db6096).

Let me raise this concern and see what the best way to deal with these type of workflows.

**Comment by febuiles** [comment]
@marcransome I reached out to the team, and although this is not expected behavior, it's also not a feature that Dependency Graph currently plans to tackle. I feel the best we can do is to note this in the `README` until support for transitive dependencies/reusable workflows is added to the API. Do you want to add a PR to make this behavior explicit in the README/docs? If not I can get to it sometime next week.

Closing this issue because I don't think there's anything else we can do atm from the Action side.
